backend/data/processors/program_conditions_pre_processing.py

```diff
@@ -0,0 +1,157 @@
+"""
+DOCUMENTATION: TBA
+
+This file currently does the job of two. TODO: move stuff out
+
+1. Pre-process condition tokenisations
+2. Tokenise conditions
+
+"""
+
+
+import re
+from typing import Dict, List
+
+from data.utility.data_helpers import read_data, write_data
+
+PROGRAMS_PROCESSED_PATH = "data/final_data/programsProcessed.json"
+PRE_PROCESSED_DATA_PATH = "data/final_data/programsConditionsPreProcessed.json"
+
+def pre_process():
+    """
+    Start Here:
+    """
+    # Raw data of programs
+    program_info = read_data(PROGRAMS_PROCESSED_PATH)
+
+    # At this stage, conditions are pre-processed
+    pre_processed = {
+        code: filter_pre_processable_conditions(info)
+        for code, info in program_info.items()
+    }
+
+    # Only care for programs with relevant coditions
+    pre_pre_processed_shortlist = shortlist_pre_proc(pre_processed)
+    pre_processed_shortlist = {
+        program: pre_process_program_requirements(cond)
+        for program, conds in pre_pre_processed_shortlist.items()
+        for cond in conds
+    }
+    print(f"Found a total of {len(pre_pre_processed_shortlist)} relevant programs")
+    write_data(pre_processed_shortlist, PRE_PROCESSED_DATA_PATH)
+    return pre_processed_shortlist
+
+
+def pre_process_program_requirements(condition_raw: Dict[str, str]) -> List[Dict]:
+    """
+    Do epic pre-proc (i only want to take in the relevant ones)
+    If you feed me a condition with no notes, i will literally die
+
+    Currently assumes only 'maturity' conditions exist. To add support for more
+    need to actually check what the condition type is first
+    """
+    notes: str = condition_raw.get("notes", "")
+    # Assumption: Only one condition per sentence. No misc. `.`
+    notes: List[str] = [
+        note.strip() for note in notes.split(".")
+        if len(note) > 0
+    ]
+    print("NOTES:", notes)
+    return [
+        pre_process_maturity_condition(note)
+        for note in notes
+    ]
+
+def pre_process_maturity_condition(string: str) -> List[str]:
+    """
+    Pre-processes a maturity condition.
+    These conditions are constructed of two sections
+        - "dependency": This is the conditin that must be passed before anything
+                        from the dependant my be satisfied.
+        - "dependant": This is the condition that is restricted and cannot be
+                        fulfilled before the dependency is satisfied.
+    """
+    print("\n"*2)
+    # print("am dealing with maturity:", string)
+    components: List[str] = string.split("before")
+    print("components:", components)
+    dependency: str = components[0].strip()
+    dependant: str = components[1].strip()
+    print("reduction:::::", dependency, "---------", dependant)
+    return {
+        # "dependency": "", "dependant": "",
+        "dependency": dependency,
+        "dependant": dependant
+    }
+
+
+def maturity_match(string: str):
```

Comment:
are there other types of notes that can be supported similarly? If so, we should make cards for them too.

Reply:
The notes very alot but not *really* anything afaict that may lend itself to a condition.
When we get to hons, there's this interesting one that pops up a bit:

```
"notes": "1. Students may substitute up to 12 UOC of Advanced Disciplinary courses (Level 5) in place of Disciplinary Knowledge courses from a school specified stream list of courses with the approval of the Program or Stream Authority.\n2. Only courses in the BE (Hons) in Materials Science and Engineering will count towards the honours calculation."
```

```
"notes": "Students must complete either a 6 UOC Synthesis of Learning Course, or a 0 UOC Synthesis Experience.\nStudents who complete a Synthesis course that is also a Work Integrated Learning course within a major, will meet a major elective, WIL and Synthesis requirement.\nStudents must complete their Final Year Synthesis course within their final year of their Bachelor of Commerce program. That is, within 48 UOC or less remaining in their BCom program."
```

But those don't seem too suitable to become conditions. However, something like this:

```
"notes": "A maximum of 72 UOC of Level 1 courses can be taken, including any General Education or mainstream Level 1 course taken to fulfil either the General Education or the Free Elective requirement."
```
seems like it would be perfect.

As would:
```
"notes": "Students cannot complete a minor with the same name as their nominated major, and Level II and III courses cannot be double-counted between majors and minors. More than one minor may be completed subject to the limit on double-counting. Students must declare their minor(s) before their final term."
```

```
"notes": "Students may not undertake any of the following excluded courses."
```